File: DataFeedViews/OutputOptionsDataFeedView.vb
'************************************************************************************
' STSimStockFlow: A .NET library for simulating stocks and flows
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Core
Imports System.Globalization
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class OutputOptionsDataFeedView

    Const DEFAULT_TIMESTEP_VALUE As String = "1"

    Public Overrides Sub LoadDataFeed(dataFeed As DataFeed)

        MyBase.LoadDataFeed(dataFeed)

        Me.SetCheckBoxBinding(Me.CheckBoxSummaryST, DATASHEET_OO_SUMMARY_OUTPUT_ST_COLUMN_NAME)
        Me.SetTextBoxBinding(Me.TextBoxSummarySTTimesteps, DATASHEET_OO_SUMMARY_OUTPUT_ST_TIMESTEPS_COLUMN_NAME)
        Me.SetCheckBoxBinding(Me.CheckBoxSummaryFL, DATASHEET_OO_SUMMARY_OUTPUT_FL_COLUMN_NAME)
        Me.SetTextBoxBinding(Me.TextBoxSummaryFLTimesteps, DATASHEET_OO_SUMMARY_OUTPUT_FL_TIMESTEPS_COLUMN_NAME)
        Me.SetCheckBoxBinding(Me.CheckBoxSpatialST, DATASHEET_OO_SPATIAL_OUTPUT_ST_COLUMN_NAME)
        Me.SetTextBoxBinding(Me.TextBoxSpatialSTTimesteps, DATASHEET_OO_SPATIAL_OUTPUT_ST_TIMESTEPS_COLUMN_NAME)
        Me.SetCheckBoxBinding(Me.CheckBoxSpatialFL, DATASHEET_OO_SPATIAL_OUTPUT_FL_COLUMN_NAME)
        Me.SetTextBoxBinding(Me.TextBoxSpatialFLTimesteps, DATASHEET_OO_SPATIAL_OUTPUT_FL_TIMESTEPS_COLUMN_NAME)

        Me.MonitorDataSheet(
          "STSim_Terminology",
          AddressOf Me.OnTerminologyChanged,
          True)

    End Sub

    ''' <summary>
    ''' A callback for when the terminology changes
    ''' </summary>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub OnTerminologyChanged(ByVal e As DataSheetMonitorEventArgs)

        Dim t As String = CStr(e.GetValue(
            "TimestepUnits", "timestep")).ToLower(CultureInfo.CurrentCulture)

        Me.LabelSummarySTTimesteps.Text = t
        Me.LabelSummaryFLTimesteps.Text = t
        Me.LabelSpatialSTTimesteps.Text = t
        Me.LabelSpatialFLTimesteps.Text = t

    End Sub

    ''' <summary>
    ''' Overrides OnBoundCheckBoxChanged
    ''' </summary>
    ''' <param name="checkBox"></param>
    ''' <param name="columnName"></param>
    ''' <remarks></remarks>
    Protected Overrides Sub OnBoundCheckBoxChanged(checkBox As System.Windows.Forms.CheckBox, columnName As String)

        MyBase.OnBoundCheckBoxChanged(checkBox, columnName)

        If (checkBox Is Me.CheckBoxSummaryST And Me.CheckBoxSummaryST.Checked And String.IsNullOrEmpty(Me.TextBoxSummarySTTimesteps.Text)) Then
            Me.SetTextBoxData(Me.TextBoxSummarySTTimesteps, DEFAULT_TIMESTEP_VALUE)
        ElseIf (checkBox Is Me.CheckBoxSummaryFL And Me.CheckBoxSummaryFL.Checked And String.IsNullOrEmpty(Me.TextBoxSummaryFLTimesteps.Text)) Then
            Me.SetTextBoxData(Me.TextBoxSummaryFLTimesteps, DEFAULT_TIMESTEP_VALUE)
        ElseIf (checkBox Is Me.CheckBoxSpatialST And Me.CheckBoxSpatialST.Checked And String.IsNullOrEmpty(Me.TextBoxSpatialSTTimesteps.Text)) Then
            Me.SetTextBoxData(Me.TextBoxSpatialSTTimesteps, DEFAULT_TIMESTEP_VALUE)
        ElseIf (checkBox Is Me.CheckBoxSpatialFL And Me.CheckBoxSpatialFL.Checked And String.IsNullOrEmpty(Me.TextBoxSpatialFLTimesteps.Text)) Then
            Me.SetTextBoxData(Me.TextBoxSpatialFLTimesteps, DEFAULT_TIMESTEP_VALUE)
        End If

        Me.EnableControls()

    End Sub

    ''' <summary>
    ''' Overrides EnableView
    ''' </summary>
    ''' <param name="enable"></param>
    ''' <remarks></remarks>
    Public Overrides Sub EnableView(enable As Boolean)

        MyBase.EnableView(enable)

        If (enable) Then
            Me.EnableControls()
        End If

    End Sub

    ''' <summary>
    ''' Enables all controls
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub EnableControls()

        'Text Boxes
        Me.TextBoxSummarySTTimesteps.Enabled = Me.CheckBoxSummaryST.Checked
        Me.TextBoxSummaryFLTimesteps.Enabled = Me.CheckBoxSummaryFL.Checked
        Me.TextBoxSpatialSTTimesteps.Enabled = Me.CheckBoxSpatialST.Checked
        Me.TextBoxSpatialFLTimesteps.Enabled = Me.CheckBoxSpatialFL.Checked

        'Timesteps labels
        Me.LabelSummarySTTimesteps.Enabled = Me.CheckBoxSummaryST.Checked
        Me.LabelSummaryFLTimesteps.Enabled = Me.CheckBoxSummaryFL.Checked
        Me.LabelSpatialSTTimesteps.Enabled = Me.CheckBoxSpatialST.Checked
        Me.LabelSpatialFLTimesteps.Enabled = Me.CheckBoxSpatialFL.Checked

    End Sub

    ''' <summary>
    ''' Handles the Clicked event for the ClearAll button
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub ButtonClearAll_Click(sender As System.Object, e As System.EventArgs) Handles ButtonClearAll.Click

        Me.ResetBoundControls()
        Me.ClearDataSheets()
        Me.EnableControls()

    End Sub

End Class

File: src/DataFeedViews/OutputOptionsDataFeedView.cs
// stsim-stockflow: SyncroSim Add-On Package (to stsim) for integrating stocks and flows into state-and-transition simulation models in ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.Core;
using System.Globalization;
using System.Reflection;

namespace SyncroSim.STSimStockFlow
{
	[ObfuscationAttribute(Exclude=true, ApplyToMembers=false)]
	internal partial class OutputOptionsDataFeedView
	{
		public OutputOptionsDataFeedView()
		{
			InitializeComponent();
		}

		private const string DEFAULT_TIMESTEP_VALUE = "1";

		public override void LoadDataFeed(DataFeed dataFeed)
		{
			base.LoadDataFeed(dataFeed);

			this.SetCheckBoxBinding(this.CheckBoxSummaryST, Constants.DATASHEET_OO_SUMMARY_OUTPUT_ST_COLUMN_NAME);
			this.SetTextBoxBinding(this.TextBoxSummarySTTimesteps, Constants.DATASHEET_OO_SUMMARY_OUTPUT_ST_TIMESTEPS_COLUMN_NAME);
			this.SetCheckBoxBinding(this.CheckBoxSummaryFL, Constants.DATASHEET_OO_SUMMARY_OUTPUT_FL_COLUMN_NAME);
			this.SetTextBoxBinding(this.TextBoxSummaryFLTimesteps, Constants.DATASHEET_OO_SUMMARY_OUTPUT_FL_TIMESTEPS_COLUMN_NAME);
			this.SetCheckBoxBinding(this.CheckBoxSpatialST, Constants.DATASHEET_OO_SPATIAL_OUTPUT_ST_COLUMN_NAME);
			this.SetTextBoxBinding(this.TextBoxSpatialSTTimesteps, Constants.DATASHEET_OO_SPATIAL_OUTPUT_ST_TIMESTEPS_COLUMN_NAME);
			this.SetCheckBoxBinding(this.CheckBoxSpatialFL, Constants.DATASHEET_OO_SPATIAL_OUTPUT_FL_COLUMN_NAME);
			this.SetTextBoxBinding(this.TextBoxSpatialFLTimesteps, Constants.DATASHEET_OO_SPATIAL_OUTPUT_FL_TIMESTEPS_COLUMN_NAME);

			this.MonitorDataSheet("STSim_Terminology", this.OnTerminologyChanged, true);
			this.AddStandardCommands();
		}

		public override void EnableView(bool enable)
		{
			base.EnableView(enable);

			if (enable)
			{
				this.EnableControls();
			}
		}

		protected override void OnRowsAdded(object sender, DataSheetRowEventArgs e)
		{
			base.OnRowsAdded(sender, e);

			if (this.ShouldEnableView())
			{
				this.EnableControls();
			}
		}

		protected override void OnRowsDeleted(object sender, DataSheetRowEventArgs e)
		{
			base.OnRowsDeleted(sender, e);

			if (this.ShouldEnableView())
			{
				this.EnableControls();
			}
		}

		private void OnTerminologyChanged(DataSheetMonitorEventArgs e)
		{
			string t = Convert.ToString(
                e.GetValue("TimestepUnits", "timestep"), CultureInfo.InvariantCulture).
                ToLower(CultureInfo.InvariantCulture);

			this.LabelSummarySTTimesteps.Text = t;
			this.LabelSummaryFLTimesteps.Text = t;
			this.LabelSpatialSTTimesteps.Text = t;
			this.LabelSpatialFLTimesteps.Text = t;
		}

		protected override void OnBoundCheckBoxChanged(System.Windows.Forms.CheckBox checkBox, string columnName)
		{
			base.OnBoundCheckBoxChanged(checkBox, columnName);

			if (checkBox == this.CheckBoxSummaryST && this.CheckBoxSummaryST.Checked & string.IsNullOrEmpty(this.TextBoxSummarySTTimesteps.Text))
			{
				this.SetTextBoxData(this.TextBoxSummarySTTimesteps, DEFAULT_TIMESTEP_VALUE);
			}
			else if (checkBox == this.CheckBoxSummaryFL && this.CheckBoxSummaryFL.Checked & string.IsNullOrEmpty(this.TextBoxSummaryFLTimesteps.Text))
			{
				this.SetTextBoxData(this.TextBoxSummaryFLTimesteps, DEFAULT_TIMESTEP_VALUE);
			}
			else if (checkBox == this.CheckBoxSpatialST && this.CheckBoxSpatialST.Checked & string.IsNullOrEmpty(this.TextBoxSpatialSTTimesteps.Text))
			{
				this.SetTextBoxData(this.TextBoxSpatialSTTimesteps, DEFAULT_TIMESTEP_VALUE);
			}
			else if (checkBox == this.CheckBoxSpatialFL && this.CheckBoxSpatialFL.Checked & string.IsNullOrEmpty(this.TextBoxSpatialFLTimesteps.Text))
			{
				this.SetTextBoxData(this.TextBoxSpatialFLTimesteps, DEFAULT_TIMESTEP_VALUE);
			}

			this.EnableControls();
		}

		private void EnableControls()
		{
			//Text Boxes
			this.TextBoxSummarySTTimesteps.Enabled = this.CheckBoxSummaryST.Checked;
			this.TextBoxSummaryFLTimesteps.Enabled = this.CheckBoxSummaryFL.Checked;
			this.TextBoxSpatialSTTimesteps.Enabled = this.CheckBoxSpatialST.Checked;
			this.TextBoxSpatialFLTimesteps.Enabled = this.CheckBoxSpatialFL.Checked;

			//Timesteps labels
			this.LabelSummarySTTimesteps.Enabled = this.CheckBoxSummaryST.Checked;
			this.LabelSummaryFLTimesteps.Enabled = this.CheckBoxSummaryFL.Checked;
			this.LabelSpatialSTTimesteps.Enabled = this.CheckBoxSpatialST.Checked;
			this.LabelSpatialFLTimesteps.Enabled = this.CheckBoxSpatialFL.Checked;
		}
	}
}